libsignal-service/src/cipher.rs

```diff
@@ -83,84 +97,99 @@ impl ServiceCipher {
-                            .source_address()
-                            .expect("Envelope with source"),
-                        envelope.source_device(),
-                    )?,
-                    Metadata {
-                        sender: envelope
-                            .source_address()
-                            .expect("envelope with source"),
-                        sender_device: envelope.source_device(),
-                        timestamp: envelope.timestamp(),
-                        needs_receipt: false,
-                    },
-                )
-            }
-            Type::UnidentifiedSender => {
-                // is unidentified sender
-                return Err(ServiceError::InvalidFrameError {
-                    reason: "UnidentifiedSender requires SealedSessionCipher, see: https://github.com/Michael-F-Bryan/libsignal-service-rs/issues/10".into(),
-                });
-            }
-            _ => {
-                // else
-                return Err(ServiceError::InvalidFrameError {
-                    reason: "Envelope has unknown type.".into(),
-                });
-            }
-        };
-
-        let mut plaintext = match envelope.r#type() {
+        let plaintext = match envelope.r#type() {
             Type::PrekeyBundle => {
+                let sender = self.get_preferred_protocol_address(
+                    envelope.source_address().expect("Envelope with source"),
+                    envelope.source_device(),
+                )?;
+                let metadata = Metadata {
+                    sender: envelope
+                        .source_address()
+                        .expect("envelope with source"),
+                    sender_device: envelope.source_device() as i32,
+                    timestamp: envelope.timestamp(),
+                    needs_receipt: false,
+                };
                 let cipher = SessionCipher::new(
                     &self.context,
                     &self.store_context,
-                    &address,
+                    &sender,
                 )?;
-                let buf = cipher.decrypt_pre_key_message(
-                    &PreKeySignalMessage::deserialize(
-                        &self.context,
-                        ciphertext,
-                    )?,
-                )?;
-                Plaintext {
-                    metadata,
-                    data: Vec::from(buf.as_slice()),
-                }
+                let mut data = cipher
+                    .decrypt_pre_key_message(
+                        &PreKeySignalMessage::deserialize(
+                            &self.context,
+                            ciphertext,
+                        )?,
+                    )?
+                    .as_slice()
+                    .to_vec();
+                let version =
+                    self.store_context.load_session(&sender)?.state().version();
+                strip_padding(version, &mut data)?;
```

Comment:
Starting to think that `strip_padding` should accept `&'a [u8]` and return `&'a [u8]`... Then we don't need the mutability. Maybe another time...

Reply:
Maybe `strip_padding` calls should go into the decrypt functions themselves? I think that would make sense.

Reply:
I only put it there to be consistent with `libsignal-metadata-java` which returns the padded message, but it doesn't really make any sense if we keep the `SealedSessionCipher` in this crate.

If we split the crates though, it might.

Reply:
Ok, for now I would not change it just because of separation of concern, I think the `decrypt` functions should return the stream as it's intended to be.

Reply:
Postponed to #46